qiskit_experiments/database_service/db_experiment_data.py

```diff
@@ -288,80 +357,62 @@ def add_analysis_callback(self, callback: Callable, **kwargs: Any):
-                self._callback_statuses[callback_id].status = JobStatus.ERROR
-                error_msg = f"Analysis callback {callback} failed: \n" "".join(
-                    traceback.format_exception(type(ex), ex, ex.__traceback__)
-                )
-                self._callback_statuses[callback_id].error_msg = error_msg
-                LOG.warning("Analysis callback %s failed:\n%s", callback, traceback.format_exc())
-            self._callback_statuses[callback_id].event.set()
-
-        with self._job_futures.lock:
-            # Determine if a future is running that we need to add callback to
-            fut_done = True
-            if self._job_futures:
-                _, fut = self._job_futures[-1]
-                fut_done = fut.done()
-            if fut_done and self._callback_future is not None:
-                fut = self._callback_future
-                fut_done = fut.done()
-            if fut_done:
-                fut = None
-
-            if fut_done:
-                # Submit future so analysis can run async even if there are no
-                # running jobs or running analysis.
-                self._callback_future = self._executor.submit(_wrapped_callback)
+        with self._job_futures.lock and self._analysis_futures.lock:
+            # Create callback dataclass
+            cid = uuid.uuid4().hex
+            self._analysis_callbacks[cid] = AnalysisCallback(
+                name=callback.__name__,
+                callback_id=cid,
+                status=AnalysisStatus.QUEUED,
+            )
+
+            # Get futures to wait for before running analysis
+            if self._analysis_futures:
+                # If analysis futures are present wait for all previous analysis
+                # futures to complete before running the latest one
+                futs = self._analysis_futures.values()
             else:
-                # Wrap the wrapped function for the format expected by Python
-                # Future.add_done_callback
-                def _done_callback(fut):
-                    if fut.cancelled():
-                        self._callback_statuses[callback_id].status = JobStatus.CANCELLED
-                        self._callback_statuses[callback_id].event.set()
-                    else:
-                        _wrapped_callback()
-
-                fut.add_done_callback(_done_callback)
-
-    def _add_jobs_data(
-        self,
-        jobs: List[Union[Job, BaseJob]],
-        timeout: Optional[float] = None,
-    ) -> None:
-        """Wait for a job to finish and add job result data.
+                # Create a dummy future to wait for jobs to finish.
+                # This allows subsequent analysis futures to be cancelled as they
+                # will be queued by the ThreadPoolExecutor while this future
+                # is still running
+                futs = [self._analysis_executor.submit(futures.wait, self._job_futures.values())]
```

Comment:
What will happen if one only adds result data (not jobs) and trigger this method? [fs](https://docs.python.org/ja/3/library/concurrent.futures.html#concurrent.futures.wait) can be None?

Reply:
In that case it evaluates `futures.wait([])` which returns done straight away